File: surface/widgets/nativescript/components/widgets/concepts/package/AuditReport.ts
import {
  StackLayout,
  Label,
  Button,
  ScrollView,
} from '@nativescript/core';

export type AuditReportState = 'idle' | 'filtering' | 'vulnSelected';
export type AuditReportEvent =
  | { type: 'FILTER'; severity?: Severity }
  | { type: 'SELECT_VULN'; id: string }
  | { type: 'CLEAR' }
  | { type: 'DESELECT' };

export function auditReportReducer(state: AuditReportState, event: AuditReportEvent): AuditReportState {
  switch (state) {
    case 'idle':
      if (event.type === 'FILTER') return 'filtering';
      if (event.type === 'SELECT_VULN') return 'vulnSelected';
      return state;
    case 'filtering':
      if (event.type === 'CLEAR') return 'idle';
      return state;
    case 'vulnSelected':
      if (event.type === 'DESELECT') return 'idle';
      return state;
    default:
      return state;
  }
}

export type Severity = 'critical' | 'high' | 'moderate' | 'low';

export interface Vulnerability {
  id: string;
  title: string;
  severity: Severity;
  package: string;
  installedVersion: string;
  patchedVersion?: string;
  description: string;
  url?: string;
}

export interface AuditReportProps {
  vulnerabilities: Vulnerability[];
  severityCounts?: Record<Severity, number>;
  lastScan: string;
  status: string;
  filterSeverity?: Severity;
  showRemediation?: boolean;
}

const SEVERITY_ORDER: Record<Severity, number> = { critical: 0, high: 1, moderate: 2, low: 3 };
const SEVERITY_COLORS: Record<Severity, string> = {
  critical: '#dc2626',
  high: '#ea580c',
  moderate: '#ca8a04',
  low: '#2563eb',
};
const ALL_SEVERITIES: Severity[] = ['critical', 'high', 'moderate', 'low'];

function countBySeverity(vulns: Vulnerability[]): Record<Severity, number> {
  const counts: Record<Severity, number> = { critical: 0, high: 0, moderate: 0, low: 0 };
  for (const v of vulns) {
    if (v.severity in counts) counts[v.severity]++;
  }
  return counts;
}

export function createAuditReport(props: AuditReportProps): {
  view: StackLayout;
  dispose: () => void;
} {
  let state: AuditReportState = props.filterSeverity ? 'filtering' : 'idle';
  let activeFilter: Severity | undefined = props.filterSeverity;
  let selectedId: string | null = null;
  const disposers: (() => void)[] = [];

  function send(event: AuditReportEvent) {
    state = auditReportReducer(state, event);
    update();
  }

  const root = new StackLayout();
  root.className = 'audit-report';
  root.automationText = 'Security audit report';

  const header = new StackLayout();
  header.orientation = 'horizontal';
  const statusLbl = new Label();
  statusLbl.text = props.status;
  statusLbl.fontWeight = 'bold';
  header.addChild(statusLbl);
  const scanLbl = new Label();
  scanLbl.text = `Last scan: ${props.lastScan}`;
  scanLbl.marginLeft = 12;
  header.addChild(scanLbl);
  root.addChild(header);

  const distBar = new StackLayout();
  distBar.orientation = 'horizontal';
  distBar.height = 8;
  root.addChild(distBar);

  const filterRow = new StackLayout();
  filterRow.orientation = 'horizontal';
  filterRow.marginTop = 8;
  filterRow.marginBottom = 8;
  root.addChild(filterRow);

  const vulnScroll = new ScrollView();
  const vulnList = new StackLayout();
  vulnScroll.content = vulnList;
  root.addChild(vulnScroll);

  function update() {
    const counts = props.severityCounts ?? countBySeverity(props.vulnerabilities);
    const total = Object.values(counts).reduce((s, n) => s + n, 0);

    distBar.removeChildren();
    for (const sev of ALL_SEVERITIES) {
      if (counts[sev] > 0) {
        const seg = new StackLayout();
        seg.backgroundColor = SEVERITY_COLORS[sev];
        seg.height = 8;
        distBar.addChild(seg);
      }
    }

    filterRow.removeChildren();
    for (const sev of ALL_SEVERITIES) {
      const btn = new Button();
      btn.text = `${counts[sev]} ${sev}`;
      btn.className = activeFilter === sev && state === 'filtering' ? 'severity-active' : 'severity-btn';
      btn.on('tap', () => {
        if (activeFilter === sev && state === 'filtering') {
          activeFilter = undefined;
          send({ type: 'CLEAR' });
        } else {
          activeFilter = sev;
          send({ type: 'FILTER', severity: sev });
        }
      });
      filterRow.addChild(btn);
    }

    const sorted = [...props.vulnerabilities].sort(
      (a, b) => SEVERITY_ORDER[a.severity] - SEVERITY_ORDER[b.severity],
    );
    const displayed =
      state === 'filtering' && activeFilter
        ? sorted.filter((v) => v.severity === activeFilter)
        : sorted;

    vulnList.removeChildren();
    if (displayed.length === 0) {
      const empty = new Label();
      empty.text =
        state === 'filtering'
          ? 'No vulnerabilities match the selected severity.'
          : 'No vulnerabilities found.';
      empty.textAlignment = 'center';
      empty.padding = 16;
      vulnList.addChild(empty);
      return;
    }

    for (const vuln of displayed) {
      const isExpanded = selectedId === vuln.id && state === 'vulnSelected';
      const item = new StackLayout();
      item.borderWidth = 1;
      item.borderColor = isExpanded ? SEVERITY_COLORS[vuln.severity] : '#e5e7eb';
      item.borderLeftWidth = 4;
      item.borderLeftColor = SEVERITY_COLORS[vuln.severity];
      item.borderRadius = 6;
      item.padding = 12;
      item.marginBottom = 8;

      const summaryRow = new StackLayout();
      summaryRow.orientation = 'horizontal';

      const sevBadge = new Label();
      sevBadge.text = vuln.severity.toUpperCase();
      sevBadge.color = '#ffffff' as any;
      sevBadge.backgroundColor = SEVERITY_COLORS[vuln.severity];
      sevBadge.padding = '2 8';
      sevBadge.borderRadius = 9999;
      sevBadge.fontSize = 12;
      sevBadge.fontWeight = 'bold';
      summaryRow.addChild(sevBadge);

      const titleLbl = new Label();
      titleLbl.text = vuln.title;
      titleLbl.fontWeight = 'bold';
      titleLbl.marginLeft = 8;
      summaryRow.addChild(titleLbl);

      const pkgLbl = new Label();
      pkgLbl.text = `${vuln.package}@${vuln.installedVersion}`;
      pkgLbl.fontSize = 13;
      pkgLbl.marginLeft = 8;
      pkgLbl.fontFamily = 'monospace';
      summaryRow.addChild(pkgLbl);

      item.addChild(summaryRow);

      if (isExpanded) {
        const descLbl = new Label();
        descLbl.text = vuln.description;
        descLbl.textWrap = true;
        descLbl.marginTop = 8;
        item.addChild(descLbl);

        if (props.showRemediation !== false) {
          const remLbl = new Label();
          if (vuln.patchedVersion) {
            remLbl.text = `Fix: Upgrade ${vuln.package} to ${vuln.patchedVersion}`;
          } else {
            remLbl.text = 'No patch available';
            remLbl.color = '#b91c1c' as any;
          }
          remLbl.fontSize = 13;
          remLbl.marginTop = 4;
          item.addChild(remLbl);
        }

        if (vuln.url) {
          const linkLbl = new Label();
          linkLbl.text = 'View advisory';
          linkLbl.color = '#2563eb' as any;
          linkLbl.fontSize = 13;
          linkLbl.marginTop = 4;
          item.addChild(linkLbl);
        }
      }

      item.on('tap', () => {
        if (selectedId === vuln.id) {
          selectedId = null;
          send({ type: 'DESELECT' });
        } else {
          selectedId = vuln.id;
          send({ type: 'SELECT_VULN', id: vuln.id });
        }
      });

      vulnList.addChild(item);
    }
  }

  update();

  return {
    view: root,
    dispose() {
      disposers.forEach((d) => d());
    },
  };
}

export default createAuditReport;

File: surface/widgets/compose/components/widgets/concepts/package/AuditReport.kt
package com.clef.surface.widgets.concepts.pkg

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// --- State machine ---

enum class AuditReportState { Idle, Filtering, VulnSelected }

sealed class AuditReportEvent {
    data class Filter(val severity: String) : AuditReportEvent()
    data class SelectVuln(val id: String) : AuditReportEvent()
    object Clear : AuditReportEvent()
    object Deselect : AuditReportEvent()
}

fun auditReportReduce(state: AuditReportState, event: AuditReportEvent): AuditReportState = when (state) {
    AuditReportState.Idle -> when (event) {
        is AuditReportEvent.Filter -> AuditReportState.Filtering
        is AuditReportEvent.SelectVuln -> AuditReportState.VulnSelected
        else -> state
    }
    AuditReportState.Filtering -> when (event) {
        is AuditReportEvent.Clear -> AuditReportState.Idle
        else -> state
    }
    AuditReportState.VulnSelected -> when (event) {
        is AuditReportEvent.Deselect -> AuditReportState.Idle
        else -> state
    }
}

// --- Types ---

data class Vulnerability(
    val id: String,
    val title: String,
    val severity: String,
    val packageName: String,
    val installedVersion: String,
    val patchedVersion: String? = null,
    val description: String,
    val url: String? = null
)

// --- Helpers ---

private val SEVERITY_ORDER = mapOf("critical" to 0, "high" to 1, "moderate" to 2, "low" to 3)
private val SEVERITY_COLORS = mapOf(
    "critical" to Color(0xFFDC2626), "high" to Color(0xFFEA580C),
    "moderate" to Color(0xFFCA8A04), "low" to Color(0xFF2563EB)
)
private val ALL_SEVERITIES = listOf("critical", "high", "moderate", "low")

private fun countBySeverity(vulns: List<Vulnerability>): Map<String, Int> {
    val counts = mutableMapOf("critical" to 0, "high" to 0, "moderate" to 0, "low" to 0)
    for (v in vulns) counts[v.severity] = (counts[v.severity] ?: 0) + 1
    return counts
}

@Composable
fun AuditReport(
    vulnerabilities: List<Vulnerability>,
    lastScan: String,
    status: String,
    modifier: Modifier = Modifier,
    showRemediation: Boolean = true
) {
    var state by remember { mutableStateOf(AuditReportState.Idle) }
    var activeFilter by remember { mutableStateOf<String?>(null) }
    var selectedId by remember { mutableStateOf<String?>(null) }

    val counts = remember(vulnerabilities) { countBySeverity(vulnerabilities) }
    val sorted = remember(vulnerabilities) {
        vulnerabilities.sortedBy { SEVERITY_ORDER[it.severity] ?: 4 }
    }
    val displayed = remember(sorted, state, activeFilter) {
        if (state == AuditReportState.Filtering && activeFilter != null) sorted.filter { it.severity == activeFilter }
        else sorted
    }

    Column(modifier = modifier.semantics { contentDescription = "Security audit report" }) {
        // Header
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            modifier = Modifier.padding(bottom = 8.dp)
        ) {
            Text(status, fontWeight = FontWeight.SemiBold, fontSize = 14.sp)
            Spacer(Modifier.weight(1f))
            Text("Last scan: $lastScan", fontSize = 12.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }

        // Severity distribution bar
        val total = counts.values.sum()
        if (total > 0) {
            Row(
                modifier = Modifier.fillMaxWidth().height(8.dp)
            ) {
                ALL_SEVERITIES.forEach { sev ->
                    val count = counts[sev] ?: 0
                    if (count > 0) {
                        Box(
                            modifier = Modifier
                                .weight(count.toFloat() / total)
                                .fillMaxHeight()
                                .then(Modifier.semantics { contentDescription = "$sev: $count" })
                        ) {
                            Surface(color = SEVERITY_COLORS[sev] ?: Color.Gray, modifier = Modifier.fillMaxSize()) {}
                        }
                    }
                }
            }
        }

        // Severity filter badges
        Row(
            horizontalArrangement = Arrangement.spacedBy(4.dp),
            modifier = Modifier.padding(vertical = 8.dp)
        ) {
            ALL_SEVERITIES.forEach { sev ->
                val isActive = activeFilter == sev && state == AuditReportState.Filtering
                FilterChip(
                    selected = isActive,
                    onClick = {
                        if (isActive) {
                            activeFilter = null
                            state = auditReportReduce(state, AuditReportEvent.Clear)
                        } else {
                            activeFilter = sev
                            state = auditReportReduce(state, AuditReportEvent.Filter(sev))
                        }
                    },
                    label = { Text("${counts[sev] ?: 0} $sev", fontSize = 11.sp) }
                )
            }
        }

        // Vulnerability list
        LazyColumn(modifier = Modifier.weight(1f)) {
            itemsIndexed(displayed) { _, vuln ->
                val isExpanded = selectedId == vuln.id && state == AuditReportState.VulnSelected
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 2.dp)
                        .clickable {
                            if (selectedId == vuln.id) {
                                selectedId = null
                                state = auditReportReduce(state, AuditReportEvent.Deselect)
                            } else {
                                selectedId = vuln.id
                                state = auditReportReduce(state, AuditReportEvent.SelectVuln(vuln.id))
                            }
                        },
                    colors = CardDefaults.cardColors(
                        containerColor = if (isExpanded) MaterialTheme.colorScheme.errorContainer
                        else MaterialTheme.colorScheme.surface
                    )
                ) {
                    Column(Modifier.padding(8.dp)) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(6.dp)
                        ) {
                            Surface(color = SEVERITY_COLORS[vuln.severity] ?: Color.Gray, shape = MaterialTheme.shapes.small) {
                                Text(
                                    vuln.severity.uppercase(),
                                    fontSize = 10.sp, fontWeight = FontWeight.Bold,
                                    color = Color.White,
                                    modifier = Modifier.padding(horizontal = 6.dp, vertical = 1.dp)
                                )
                            }
                            Text(vuln.title, fontWeight = FontWeight.SemiBold, fontSize = 13.sp, modifier = Modifier.weight(1f))
                            Text(
                                "${vuln.packageName}@${vuln.installedVersion}",
                                fontSize = 11.sp, fontFamily = FontFamily.Monospace,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }

                        if (isExpanded) {
                            Text(vuln.description, fontSize = 13.sp, modifier = Modifier.padding(top = 8.dp))
                            if (showRemediation) {
                                Text(
                                    if (vuln.patchedVersion != null) "Fix: Upgrade ${vuln.packageName} to ${vuln.patchedVersion}"
                                    else "No patch available",
                                    fontSize = 12.sp,
                                    color = if (vuln.patchedVersion != null) MaterialTheme.colorScheme.onSurface else Color(0xFFB91C1C),
                                    modifier = Modifier.padding(top = 4.dp)
                                )
                            }
                        }
                    }
                }
            }

            if (displayed.isEmpty()) {
                item {
                    Text(
                        if (state == AuditReportState.Filtering) "No vulnerabilities match the selected severity."
                        else "No vulnerabilities found.",
                        modifier = Modifier.padding(12.dp),
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        }
    }
}
